dll/win32/shell32/wine/shpolicy.c

```diff
@@ -43,1276 +44,62 @@
-
-  {
-    REST_ALLOWUNHASHEDWEBVIEW,
-    strExplorer,
-    "AllowUnhashedWebView",
-    SHELL_NO_POLICY
-  },
-#if (WINE_FILEVERSION_MAJOR >= 6)
-  {
-    REST_ALLOWLEGACYWEBVIEW,
-    strExplorer,
-    "AllowLegacyWebView",
-    SHELL_NO_POLICY
-  },
-#endif
-  {
-    REST_REVERTWEBVIEWSECURITY,
-    strExplorer,
-    "RevertWebViewSecurity",
-    SHELL_NO_POLICY
-  },
-  {
-    REST_INHERITCONSOLEHANDLES,
-    strExplorer,
-    "InheritConsoleHandles",
-    SHELL_NO_POLICY
-  },
-
-#if (WINE_FILEVERSION_MAJOR >= 6)
-
-#if (NTDDI_VERSION < NTDDI_LONGHORN)
-  {
-    REST_SORTMAXITEMCOUNT,
-    strExplorer,
-    "SortMaxItemCount",
-    SHELL_NO_POLICY
-  },
-#endif
-  {
-    REST_NOREMOTERECURSIVEEVENTS,
-    strExplorer,
-    "NoRemoteRecursiveEvents",
-    SHELL_NO_POLICY
-  },
-
-#endif // WINE_FILEVERSION_MAJOR
-
-  {
-    REST_NOREMOTECHANGENOTIFY,
-    strExplorer,
-    "NoRemoteChangeNotify",
-    SHELL_NO_POLICY
-  },
-
-#if (WINE_FILEVERSION_MAJOR >= 6)
-
-#if (NTDDI_VERSION < NTDDI_LONGHORN)
-  {
-    REST_NOSIMPLENETIDLIST,
-    strExplorer,
-    "NoSimpleNetIDList",
-    SHELL_NO_POLICY
-  },
-#endif
-// #if (NTDDI_VERSION < NTDDI_LONGHORN)
-// NOTE: Geoff Chappell is inacurrate here.
-  {
-    REST_NOENUMENTIRENETWORK,
-    strExplorer,
-    "NoEnumEntireNetwork",
-    SHELL_NO_POLICY
-  },
-// #endif
-#if (NTDDI_VERSION < NTDDI_LONGHORN)
-  {
-    REST_NODETAILSTHUMBNAILONNETWORK,
-    strExplorer,
-    "NoDetailsThumbnailOnNetwork",
-    SHELL_NO_POLICY
-  },
-#endif
-  {
-    REST_NOINTERNETOPENWITH,
-    strExplorer,
-    "NoInternetOpenWith",
-    SHELL_NO_POLICY
-  },
-#if (NTDDI_VERSION < NTDDI_LONGHORN)
-  {
-    REST_ALLOWLEGACYLMZBEHAVIOR,
-    strExplorer,
-    "AllowLegacyLMZBehavior",
-    SHELL_NO_POLICY
-  },
-#endif
-  {
-    REST_DONTRETRYBADNETNAME,
-    strExplorer,
-    "DontRetryBadNetName",
-    SHELL_NO_POLICY
-  },
-  {
-    REST_ALLOWFILECLSIDJUNCTIONS,
-    strExplorer,
-    "AllowFileCLSIDJunctions",
-    SHELL_NO_POLICY
-  },
-  {
-    REST_NOUPNPINSTALL,
-    strExplorer,
-    "NoUPnPInstallTask",
-    SHELL_NO_POLICY
-  },
-
-// "NormalizeLinkNetPidls" only in version 6.0 from Windows XP SP3.
-
-#if (NTDDI_VERSION >= NTDDI_LONGHORN)
-  {
-    REST_ARP_DONTGROUPPATCHES,
-    strAddRemoveProgs,
-    "DontGroupPatches",
-    SHELL_NO_POLICY
-  },
-  {
-    REST_ARP_NOCHOOSEPROGRAMSPAGE,
-    strAddRemoveProgs,
-    "NoChooseProgramsPage",
-    SHELL_NO_POLICY
-  },
-#endif
-
-// "AllowCLSIDPROGIDMapping" in Windows XP SP3 and Windows Server 2003 SP2 only.
-// Maybe in Vista+ too?
-
-#endif // WINE_FILEVERSION_MAJOR
-
-#endif // __REACTOS__
-
-/* 0x4000061 - 0x4000086 */
-  {
-    REST_NODISCONNECT,
-    strExplorer,
-    "NoDisconnect",
-    SHELL_NO_POLICY
-  },
-  {
-    REST_NOSECURITY,
-    strExplorer,
-    "NoNTSecurity",
-    SHELL_NO_POLICY
-  },
-  {
-    REST_NOFILEASSOCIATE,
-    strExplorer,
-    "NoFileAssociate",
-    SHELL_NO_POLICY
-  },
-
-#ifdef __REACTOS__
+#include "PolicyData.h"
```

Comment:
So, here's my question:
Why did you find it necessary to move this data into a separate header, and using a macro to define each entry?

Reply:
It is the principle of "separation of program and data". There is a philosophy that raw data should be separated from program code.

Reply:
We mainly need to pay attention to the program code when programming.
Some hundred lines of data is not for programming.

Reply:
OK this answers the "why moving the data into header."
Now, why using a macro, instead of just having lines:
```c
{ REST_NORUN,        "Explorer",      "NoRun"    },
...
```
for example?

Reply:
You know preprocessor is a magic.
Functional macro has flexibility.
For example, we can convert this list to a csv file. We can reorder fieids and rows.
There is flexibility.

Reply:
Sure, but this doesn't help much if this is just used in the shell32. I would understand you use the macros if you knew right now that this file were going to be used in other modules as well.

Reply:
> Do you know about documentation cost [...]

Clearly you have missed your own point.
Because the original table had more documentation about how the values changed, etc., than the new one.

Reply:
Ok, I will convert 2nd field and 3rd field into wide string literals. Are you okay?

Reply:
So yes, please change the macro a bit so that we can have something like:
```c
DEFINE_POLICY( REST_NOENTIRENETWORK,              L"Network",       L"NoEntireNetwork"                  ),  // REGSTR_VAL_NOENTIRENETWORK
```
or
```c
DEFINE_POLICY( REST_NOENTIRENETWORK,               "Network",        "NoEntireNetwork"                  ),  // REGSTR_VAL_NOENTIRENETWORK
```
(in that latter case the macro takes care of unicoding the string if necessary).

However it doesn't make sense _**at all**_ to de-stringize the values that _**are**_ strings.

Reply:
> Ok, I will convert 2nd field and 3rd field into wide string literals. Are you okay?

Works for me (see my comment that followed it). The DEFINE_POLICY macro could look like:
```c
#define DEFINE_POLICY(policy, appstr, keystr) \
    { policy, L##appstr, L##keystr }
```

Reply:
Fixed.